### PR TITLE
Improve performance of StatsCoordinator[WIP]

### PR DESCRIPTION
**Pull request by k0nserv**
Instead of making several queries per pod
make one query per "type" e.g one query
for the total number of downloads for all pods

### CODE REVIEW

File: runner/stats_coordinator.rb [hunk 1]
```diff
@@ -53,43 +89,33 @@ def update_pod pod_id, data
...
-    def pod_try pod_name, time=nil
+    def installs time = nil
+      where = ''
+      where = "WHERE sent_at >= current_date - interval '#{time}'" if time
       query = <<-SQL
-        SELECT COUNT(dependency_name)
-        FROM install
-        WHERE dependency_name = $1
-        AND pod_try = true
+        SELECT depdency_name,
```
Comment (orta) [inline runner/stats_coordinator.rb]: typo

File: runner/stats_coordinator.rb [hunk 2]
```diff
@@ -53,43 +89,33 @@ def update_pod pod_id, data
...
+               count(CASE WHEN pod_try THEN 1 ELSE null END) AS try_true,
+               count(CASE WHEN pod_try THEN null ELSE 1 END) AS try_false
+        FROM installs
+        #{where}
+        GROUP BY depdency_name
```
Comment (floere) [inline runner/stats_coordinator.rb]: typo 2

File: runner/stats_coordinator.rb [hunk 3]
```diff
@@ -53,43 +89,33 @@ def update_pod pod_id, data
...
-        AND pod_try = false
-      SQL
-      query << "AND sent_at >= current_date - interval '#{time}'" if time
-      @connection.exec(query, [pod_name])[0]["count"].to_i || 0
-    end
-
-    def target pod_name, type, time=nil
-      type_id = PRODUCT_TYPE_UTI[type]
+    def target time = nil
+      where = ''
```
Comment (floere) [inline runner/stats_coordinator.rb]: Why assign a `where` when it's overwritten right after?

File: runner/stats_coordinator.rb [hunk 4]
```diff
@@ -53,43 +89,33 @@ def update_pod pod_id, data
...
+    def installs time = nil
+      where = ''
+      where = "WHERE sent_at >= current_date - interval '#{time}'" if time
       query = <<-SQL
-        SELECT COUNT(dependency_name)
-        FROM install
-        WHERE dependency_name = $1
-        AND pod_try = true
+        SELECT dependency_name
+               count(CASE WHEN pod_try THEN 1 ELSE null END) AS try_true,
```
Comment (floere) [inline runner/stats_coordinator.rb]: Here you could use `pod_tries` and `downloads`, or `pod_try` and `download` for the aliases.

Comment (k0nserv) [inline runner/stats_coordinator.rb]: Yeah good point will change that

**Comment by k0nserv** [comment]
```
  9.990000   1.690000  11.680000 (164.815797)
```

With this code

**Comment by orta** [comment]
So, this looks great :+1:

File: runner/stats_coordinator.rb [hunk 5]
```diff
@@ -52,44 +52,112 @@ def update_pod pod_id, data
         StatsMetrics.insert(data)
       end
     end
+    
+    def total_installs name
+      installs = installs_total.try(:[], name).try(:[], 'downloads').try(:to_i)
+      installs = 0 if installs.nil?
```
Comment (segiddins) [inline runner/stats_coordinator.rb]: `installs || 0`

Comment (k0nserv) [inline runner/stats_coordinator.rb]: :+1: 

can just be 

```ruby
def total_installs name
  installs = installs_total.try(:[], name).try(:[], 'downloads').try(:to_i)
  installs || 0
end
```

then right?

File: runner/stats_coordinator.rb [hunk 6]
```diff
@@ -52,44 +52,112 @@ def update_pod pod_id, data
...
-      query << "AND sent_at >= current_date - interval '#{time}'" if time
+    def weekly_installs name
+      installs = installs_week.try(:[], name).try(:[], 'downloads').try(:to_i)
+      installs = 0 if installs.nil?
+      installs
+    end
+
+    def monthly_installs name
+      installs = installs_month.try(:[], name).try(:[], 'downloads').try(:to_i)
+      installs = 0 if installs.nil?
```
Comment (segiddins) [inline runner/stats_coordinator.rb]: Yup, for all of these

Comment (k0nserv) [inline runner/stats_coordinator.rb]: Will do

File: runner/stats_coordinator.rb (hunk 5 above)
Comment (floere) [inline runner/stats_coordinator.rb]: That's right.

**Comment by k0nserv** [comment]
Updated with fixes from review

**Comment by orta** [comment]
Alright, shipping this :+1:

**Comment by floere** [comment]
:clap: Nice! Very curious how fast it runs.

**Comment by k0nserv** [comment]
Around 150-160 seconds from the looks of it

**Comment by floere** [comment]
Locally, on your machine?

**Comment by floere** [comment]
I'm interested how fast it runs when it interacts with the Trunk DB.

**Comment by orta** [comment]
With the production analytics server, and my local postgres setup it took 155 seconds, half of that was the work refactored above, the latter half is when we get total stats data.

Gonna run against the prod db now.

**Comment by k0nserv** [comment]
Yeah it took ~160 on my machine with production analytics server

**Comment by orta** [comment]
Ran it against production, took 688 seconds, so roughly 11 minutes.

**Comment by floere** [comment]
Very nice. So roughly 24 pods a second updated (in an estimated 500 seconds).

**Comment by floere** [comment]
We could reduce the pod updating time significantly, but this is already a big win 😊

**Comment by k0nserv** [comment]
Premature optimization ;) But yeah we could probably do that with `COPY`? Or were you thinking of some other way?

**Pull request merged**